File: les/interaction_graph.cpp
/*
 * This file implements interaction_graph.hpp interface
 *
 * Copyright (c) 2012 Alexander Sviridenko
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.  See the License for the specific language governing
 * permissions and limitations under the License.
 */

#include <les/interaction_graph.hpp>
#include <les/packed_vector.hpp>

#include <boost/graph/adjacency_list.hpp>
#include <boost/graph/connected_components.hpp>

using namespace boost;

map< int, vector<int> >
InteractionGraph::get_connected_components()
{
  map< int, vector<int> > components;
#if 0
  vector<int> vertices_to_components(get_num_vertices());
  int num = boost::connected_components(*this, &vertices_to_components[0]);
  for (size_t i = 0; i < vertices_to_components.size(); i++) {
    components[ vertices_to_components[i] ].push_back(i);
  }
#endif
#if 0
  printf("Total number of components: %d\n", num);
  for (vector<int>::size_type i = 0; i != components.size(); ++i)
    printf("Vertex %d is in component %d\n", i, components[i]);
  printf("\n");
#endif

  return components;
}

InteractionGraph::InteractionGraph(MILPP* problem) : Graph(problem->get_num_cols())
{
  int ri; /* row index */
  int ci; /* col index */
  int ni; /* neighbor index */
  PackedVector* row;
  problem_ = problem;
  /* Start forming variable neighbourhoods by using constraints. */
  for (ri = 0; ri < problem->get_num_rows(); ri++) {
    row = problem->get_row(ri);
    for (ci = 0; ci < row->get_num_elements(); ci++) {
      for (ni = 0; ni < row->get_num_elements(); ni++) {
        if ((ci == ni) ||
            has_edge(row->get_index_by_pos(ci), row->get_index_by_pos(ni)))
          continue;
        add_edge(row->get_index_by_pos(ci),
                 row->get_index_by_pos(ni));
      }
    }
  }
}

File: les/interaction_graph.py
# Copyright (c) 2012-2013 Oleksandr Sviridenko
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""The interaction graph of the discrete optimization problem (DOP) is called an
undirected graph :math:`G=(X,E)`, such that:

1. Vertices :math:`X` of :math:`G` correspond to variables of the DOP;
2. Two vertices of :math:`G` are adjacent if corresponding variables interact.

Interaction graph of variables is also called constraint graph.
"""

import networkx
import itertools

from les.problems.problem_base import ProblemBase

def _extract_indices(m, i):
  start = m.indptr[i]
  size = m.indptr[i + 1] - start
  result = []
  for j in xrange(start, start + size):
    result.append(m.indices[j])
  return result

class InteractionGraph(networkx.Graph):
  """This class represents an interaction graph of a given problem."""

  def __init__(self, problem=None):
    networkx.Graph.__init__(self)
    self._problem = None
    if problem:
      self.read_problem(problem)

  def read_problem(self, problem):
    if not isinstance(problem, ProblemBase):
      raise TypeError()
    # TODO: improve this
    for p in xrange(problem.get_num_rows()):
      J = _extract_indices(problem.get_cons_matrix(), p)
      for i in xrange(0, len(J)):
        for j in xrange(i, len(J)):
          self.add_edge(J[i], J[j])
    self._problem = problem

  def get_problem(self):
    return self._problem
